File: clients/python/peops_sdk/telemetry.py
"""Background telemetry reporter — fault-tolerant by construction.

Design contract: NOTHING in this module may ever raise into the caller's
serving path. Events queue into a bounded deque (drop-oldest under pressure,
with the drop count itself reported); a daemon thread flushes batches to
POST /api/v1/telemetry/{deployment_id}/batch with backoff; atexit performs a
final best-effort flush with a small time budget.

Disable entirely with report_telemetry=False or PEOPS_SDK_TELEMETRY=0.
"""

from __future__ import annotations

import atexit
import os
import threading
import time
import uuid
from collections import deque
from datetime import datetime, timezone
from typing import Any

from ._http import HttpSession
from .stats import WindowAggregator
from .system import runtime_fingerprint, system_sample

def _env_float(name: str, default: float) -> float:
    try:
        return float(os.environ.get(name, default))
    except ValueError:
        return default


_QUEUE_MAX = 10_000
_BATCH_MAX = 450            # below the server's 500-item cap
_FLUSH_INTERVAL_S = _env_float("PEOPS_SDK_FLUSH_INTERVAL_S", 5.0)
_SNAPSHOT_INTERVAL_S = _env_float("PEOPS_SDK_SNAPSHOT_INTERVAL_S", 30.0)
_WINDOW_INTERVAL_S = _env_float("PEOPS_SDK_WINDOW_INTERVAL_S", 60.0)
_WINDOW_MAX_REQUESTS = int(_env_float("PEOPS_SDK_WINDOW_MAX_REQUESTS", 200))
_ATEXIT_BUDGET_S = 3.0


def _iso_now() -> str:
    return datetime.now(timezone.utc).strftime("%Y-%m-%dT%H:%M:%S.%f")[:-3] + "Z"


def telemetry_enabled(flag: bool | None = None) -> bool:
    if flag is not None and not flag:
        return False
    return os.environ.get("PEOPS_SDK_TELEMETRY", "1") not in ("0", "false", "no")


class TelemetryReporter:
    """Collects events/snapshots/windows and ships them in the background."""

    def __init__(
        self,
        base_url: str,
        deployment_id: str,
        api_key: str,
        *,
        sdk_version: str,
        enabled: bool = True,
    ) -> None:
        self.enabled = telemetry_enabled(enabled)
        self.client_id = f"sdk_{uuid.uuid4().hex[:10]}"
        self.deployment_id = deployment_id
        self._events: deque[dict] = deque(maxlen=_QUEUE_MAX)
        self._snapshots: deque[dict] = deque(maxlen=64)
        self._windows: deque[dict] = deque(maxlen=64)
        self._dropped = 0
        self._sent_events = 0
        self._lock = threading.Lock()
        self._window_lock = threading.Lock()
        self._aggregator = WindowAggregator()
        self._window_requests = 0
        self._fingerprint = runtime_fingerprint(sdk_version)
        self._closed = threading.Event()
        self._thread: threading.Thread | None = None
        self._http: HttpSession | None = None
        self._throughput_marker = (time.monotonic(), 0)

        if self.enabled:
            self._http = HttpSession(
                base_url, api_key, timeout=10.0, max_attempts=2, max_backoff=30.0)
            self._thread = threading.Thread(
                target=self._loop, name="peops-telemetry", daemon=True)
            self._thread.start()
            atexit.register(self.close)

    # ── recording (hot path — must be cheap and never raise) ────────────────

    def record_event(
        self,
        *,
        latency_ms: float,
        pre_ms: float | None = None,
        post_ms: float | None = None,
        success: bool = True,
        error_code: str | None = None,
        batch_size: int = 1,
        region: str = "local",
        input_sig: str | None = None,
    ) -> None:
        if not self.enabled:
            return
        try:
            event = {
                "ts": _iso_now(),
                "latencyMs": round(float(latency_ms), 3),
                "success": success,
                "batchSize": int(batch_size),
                "region": region,
            }
            if pre_ms is not None:
                event["preMs"] = round(float(pre_ms), 3)
            if post_ms is not None:
                event["postMs"] = round(float(post_ms), 3)
            if error_code:
                event["errorCode"] = error_code
            if input_sig:
                event["inputSig"] = input_sig
            with self._lock:
                if len(self._events) == self._events.maxlen:
                    self._dropped += 1
                self._events.append(event)
                self._sent_events += 1
                self._window_requests += 1
                cut_window = self._window_requests >= _WINDOW_MAX_REQUESTS
            if cut_window:
                # Request-driven window cut: keeps window sizes deterministic
                # under burst load instead of waiting for the next loop tick.
                self._take_window()
        except Exception:
            pass

    def observe(self, inputs: dict[str, Any] | None, output: Any | None) -> None:
        if not self.enabled:
            return
        try:
            self._aggregator.observe(inputs, output)
        except Exception:
            pass

    # ── background loop ──────────────────────────────────────────────────────

    def _loop(self) -> None:
        last_snapshot = last_window = time.monotonic()
        while not self._closed.wait(_FLUSH_INTERVAL_S):
            now = time.monotonic()
            try:
                if now - last_snapshot >= _SNAPSHOT_INTERVAL_S:
                    self._take_snapshot()
                    last_snapshot = now
                window_due = (
                    now - last_window >= _WINDOW_INTERVAL_S
                    or self._window_requests >= _WINDOW_MAX_REQUESTS
                )
                if window_due:
                    self._take_window()
                    last_window = now
                self._flush()
            except Exception:
                pass  # the loop must survive anything

    def _take_snapshot(self) -> None:
        now = time.monotonic()
        marker_t, marker_n = self._throughput_marker
        with self._lock:
            sent = self._sent_events
            dropped = self._dropped
        elapsed_min = max(1e-6, (now - marker_t) / 60.0)
        rpm = (sent - marker_n) / elapsed_min
        self._throughput_marker = (now, sent)
        self._snapshots.append({
            "ts": _iso_now(),
            **system_sample(),
            "throughputRpm": round(rpm, 2),
            "droppedEvents": dropped,
            **self._fingerprint,
        })

    def _take_window(self) -> None:
        # Serialized: callable from both the hot path (request-cap cut) and
        # the background loop (time-based cut).
        with self._window_lock:
            with self._lock:
                self._window_requests = 0
            window = self._aggregator.flush()
            if window:
                self._windows.append(window)

    def _flush(self) -> bool:
        if self._http is None:
            return True
        with self._lock:
            events = [self._events.popleft()
                      for _ in range(min(_BATCH_MAX, len(self._events)))]
        snapshots = [self._snapshots.popleft() for _ in range(len(self._snapshots))]
        windows = [self._windows.popleft() for _ in range(len(self._windows))]
        if not events and not snapshots and not windows:
            return True
        try:
            self._http.request(
                "POST", f"/api/v1/telemetry/{self.deployment_id}/batch",
                json={
                    "clientId": self.client_id,
                    "events": events,
                    "snapshots": snapshots,
                    "windows": windows,
                },
            )
            return True
        except Exception:
            # Re-queue at the FRONT so ordering survives one failed flush;
            # the bounded deque drops oldest under sustained failure.
            with self._lock:
                for ev in reversed(events):
                    if len(self._events) == self._events.maxlen:
                        self._dropped += 1
                        break
                    self._events.appendleft(ev)
            for sn in reversed(snapshots):
                self._snapshots.appendleft(sn)
            for w in reversed(windows):
                self._windows.appendleft(w)
            return False

    # ── shutdown ─────────────────────────────────────────────────────────────

    def close(self) -> None:
        """Final best-effort flush within a small budget; idempotent."""
        if not self.enabled or self._closed.is_set():
            return
        self._closed.set()
        if self._thread is not None:
            self._thread.join(timeout=1.0)
        try:
            # Always ship at least one snapshot per session — it carries the
            # runtime fingerprint the dashboard's client-hosts table shows
            # (short sessions would otherwise never hit the 30s cadence).
            self._take_snapshot()
            self._take_window()
            deadline = time.monotonic() + _ATEXIT_BUDGET_S
            while time.monotonic() < deadline:
                if self._flush() and not self._events:
                    break
        except Exception:
            pass
        finally:
            if self._http is not None:
                self._http.close()

File: clients/node/src/telemetry.ts
// Background telemetry reporter — fault-tolerant by construction.
// Mirrors clients/python/astra_sdk/telemetry.py.
//
// Design contract: NOTHING here may ever throw into the caller's serving path.
// Events buffer into a bounded array (drop-oldest under pressure, with the drop
// count itself reported); an unref'd interval timer flushes batches to
// POST /api/v1/telemetry/{deployment_id}/batch with backoff; a `beforeExit` hook
// performs a final best-effort flush within a small budget.
//
// Disable entirely with enabled:false or ASTRA_SDK_TELEMETRY=0.

import { randomBytes } from "node:crypto";

import { HttpSession } from "./http.js";
import { type Sampleable, WindowAggregator } from "./stats.js";
import { type OrtRuntimeInfo, runtimeFingerprint, systemSample } from "./system.js";

function envFloat(name: string, def: number): number {
  const v = process.env[name];
  if (v === undefined) return def;
  const n = Number(v);
  return Number.isFinite(n) ? n : def;
}

const QUEUE_MAX = 10_000;
const BATCH_MAX = 450; // below the server's 500-item cap
const FLUSH_INTERVAL_S = envFloat("ASTRA_SDK_FLUSH_INTERVAL_S", 5);
const SNAPSHOT_INTERVAL_S = envFloat("ASTRA_SDK_SNAPSHOT_INTERVAL_S", 30);
const WINDOW_INTERVAL_S = envFloat("ASTRA_SDK_WINDOW_INTERVAL_S", 60);
const WINDOW_MAX_REQUESTS = Math.trunc(envFloat("ASTRA_SDK_WINDOW_MAX_REQUESTS", 200));
const ATEXIT_BUDGET_MS = 3000;

const round = (x: number, p: number): number => {
  const f = 10 ** p;
  return Math.round(x * f) / f;
};

export function telemetryEnabled(flag?: boolean): boolean {
  if (flag === false) return false;
  const v = process.env.ASTRA_SDK_TELEMETRY;
  return !(v === "0" || v === "false" || v === "no");
}

export interface TelemetryReporterOptions {
  sdkVersion: string;
  enabled?: boolean;
  activeProvider?: string;
  ortVersion?: string;
  availableProviders?: string[];
}

export interface RecordEventOptions {
  latencyMs: number;
  preMs?: number;
  postMs?: number;
  success?: boolean;
  errorCode?: string;
  batchSize?: number;
  region?: string;
  inputSig?: string;
}

export class TelemetryReporter {
  readonly enabled: boolean;
  private readonly clientId: string;
  private readonly deploymentId: string;
  private readonly fingerprint: Record<string, unknown>;
  private readonly aggregator = new WindowAggregator();

  private events: Record<string, unknown>[] = [];
  private snapshots: Record<string, unknown>[] = [];
  private windows: Record<string, unknown>[] = [];
  private dropped = 0;
  private sentEvents = 0;
  private windowRequests = 0;

  private http: HttpSession | null = null;
  private timer: ReturnType<typeof setInterval> | null = null;
  private beforeExitHandler: (() => void) | null = null;
  private closed = false;
  private flushing = false;
  private lastSnapshot = Date.now();
  private lastWindow = Date.now();
  private throughputMarkerTime = Date.now();
  private throughputMarkerN = 0;

  constructor(
    baseUrl: string,
    deploymentId: string,
    apiKey: string,
    opts: TelemetryReporterOptions,
  ) {
    this.enabled = telemetryEnabled(opts.enabled);
    this.clientId = `sdk_${randomBytes(5).toString("hex")}`;
    this.deploymentId = deploymentId;
    const ort: OrtRuntimeInfo = {
      ortVersion: opts.ortVersion,
      availableProviders: opts.availableProviders,
      activeProvider: opts.activeProvider,
    };
    this.fingerprint = runtimeFingerprint(opts.sdkVersion, ort);

    if (this.enabled) {
      this.http = new HttpSession(baseUrl, apiKey, {
        timeout: 10,
        maxAttempts: 2,
        maxBackoff: 30,
      });
      this.timer = setInterval(() => {
        void this.tick();
      }, FLUSH_INTERVAL_S * 1000);
      this.timer.unref();
      this.beforeExitHandler = () => {
        void this.close();
      };
      process.once("beforeExit", this.beforeExitHandler);
    }
  }

  // ── recording (hot path — must be cheap and never raise) ──────────────────

  recordEvent(opts: RecordEventOptions): void {
    if (!this.enabled) return;
    try {
      const event: Record<string, unknown> = {
        ts: new Date().toISOString(),
        latencyMs: round(opts.latencyMs, 3),
        success: opts.success ?? true,
        batchSize: Math.trunc(opts.batchSize ?? 1),
        region: opts.region ?? "local",
      };
      if (opts.preMs !== undefined) event.preMs = round(opts.preMs, 3);
      if (opts.postMs !== undefined) event.postMs = round(opts.postMs, 3);
      if (opts.errorCode) event.errorCode = opts.errorCode;
      if (opts.inputSig) event.inputSig = opts.inputSig;

      if (this.events.length >= QUEUE_MAX) {
        this.events.shift();
        this.dropped += 1;
      }
      this.events.push(event);
      this.sentEvents += 1;
      this.windowRequests += 1;
      if (this.windowRequests >= WINDOW_MAX_REQUESTS) this.takeWindow();
    } catch {
      /* telemetry must never break serving */
    }
  }

  observe(inputs: Record<string, Sampleable> | null | undefined, output: Sampleable | null): void {
    if (!this.enabled) return;
    try {
      this.aggregator.observe(inputs, output);
    } catch {
      /* ignore */
    }
  }

  // ── background loop ───────────────────────────────────────────────────────

  private async tick(): Promise<void> {
    if (this.closed || this.flushing) return;
    this.flushing = true;
    try {
      const now = Date.now();
      if (now - this.lastSnapshot >= SNAPSHOT_INTERVAL_S * 1000) {
        this.takeSnapshot();
        this.lastSnapshot = now;
      }
      if (
        now - this.lastWindow >= WINDOW_INTERVAL_S * 1000 ||
        this.windowRequests >= WINDOW_MAX_REQUESTS
      ) {
        this.takeWindow();
        this.lastWindow = now;
      }
      await this.flush();
    } catch {
      /* the loop must survive anything */
    } finally {
      this.flushing = false;
    }
  }

  private takeSnapshot(): void {
    const now = Date.now();
    const elapsedMin = Math.max(1e-6, (now - this.throughputMarkerTime) / 60000);
    const rpm = (this.sentEvents - this.throughputMarkerN) / elapsedMin;
    this.throughputMarkerTime = now;
    this.throughputMarkerN = this.sentEvents;
    this.snapshots.push({
      ts: new Date().toISOString(),
      ...systemSample(),
      throughputRpm: round(rpm, 2),
      droppedEvents: this.dropped,
      ...this.fingerprint,
    });
    if (this.snapshots.length > 64) this.snapshots.shift();
  }

  private takeWindow(): void {
    this.windowRequests = 0;
    const w = this.aggregator.flush();
    if (w) {
      this.windows.push(w);
      if (this.windows.length > 64) this.windows.shift();
    }
  }

  private async flush(): Promise<boolean> {
    if (!this.http) return true;
    const events = this.events.splice(0, Math.min(BATCH_MAX, this.events.length));
    const snapshots = this.snapshots.splice(0, this.snapshots.length);
    const windows = this.windows.splice(0, this.windows.length);
    if (!events.length && !snapshots.length && !windows.length) return true;
    try {
      await this.http.request("POST", `/api/v1/telemetry/${this.deploymentId}/batch`, {
        json: { clientId: this.clientId, events, snapshots, windows },
      });
      return true;
    } catch {
      // Re-queue at the FRONT so ordering survives one failed flush; the bounded
      // queue drops oldest under sustained failure.
      for (let i = events.length - 1; i >= 0; i--) {
        if (this.events.length >= QUEUE_MAX) {
          this.dropped += 1;
          break;
        }
        this.events.unshift(events[i]!);
      }
      this.snapshots.unshift(...snapshots);
      this.windows.unshift(...windows);
      return false;
    }
  }

  // ── shutdown ──────────────────────────────────────────────────────────────

  /** Final best-effort flush within a small budget; idempotent. */
  async close(): Promise<void> {
    if (!this.enabled || this.closed) return;
    this.closed = true;
    if (this.timer) {
      clearInterval(this.timer);
      this.timer = null;
    }
    if (this.beforeExitHandler) {
      process.removeListener("beforeExit", this.beforeExitHandler);
      this.beforeExitHandler = null;
    }
    try {
      // Always ship at least one snapshot per session — it carries the runtime
      // fingerprint the dashboard's client-hosts table shows (short sessions
      // would otherwise never hit the 30s cadence).
      this.takeSnapshot();
      this.takeWindow();
      const deadline = Date.now() + ATEXIT_BUDGET_MS;
      while (Date.now() < deadline) {
        const ok = await this.flush();
        if (ok && this.events.length === 0) break;
      }
    } catch {
      /* best effort */
    } finally {
      this.http?.close();
    }
  }
}
